docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+# ADR 011: kURL Add-on Operator
+
+## Context
+
+kURL today is a set of bash scripts that configure and install a Kubernetes cluster in addition to a set of add-on components. The components are installed in sequence by applying highly customized Kubernetes YAML resources. The current architecture makes kURL hard to develop, maintain, and support.
+
+As an amalgam of bash scripts, the code base suffers from low cohesion and high complexity which results in the code being difficult to read and reason, and difficult to test. The installation is applied sequentially and each component is highly dependent on each other with no scopes to limit access and enforce separation. The project is architecturally a monolith without any functional separation of components and concerns.
+
+The installation of all components is done procedurally in the script. When there is a failure, it is difficult to understand what components have failed and how that affects the installation process as a whole.
+
+The components are installed in a proprietary fashion by applying highly customized Kubernetes YAML resources. They are hard to customize as the spec is proprietary and hard to extend because each customization must be passed through configuration to the end-user. Changes to this spec must be documented and are less discoverable. This method is not supported by the upstream which makes it difficult to pin addon installation issues on the upstream project and difficult to request support from the upstream.
+
+
+## Decision
+
+We will develop a Kubernetes [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) in _Go_ that will install, configure and manage add-on applications in the cluster. The operator will extend the Kubernetes API through Custom Resource Definitions (CRD) and will allow users to declaritively manage add-ons with a Custom Resource (CR). 
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
```

Comment:
This is another point that for remains totally uncertain. It is not as simple as it may initially appear. We need to address several questions to ensure that users can pass ANY Helm chart value and that we can effectively handle them:

- Are we suggesting that users input the values through a value.yaml file? If so, we need to define the process for providing and processing this file. How that would work within the Operator?
- What happens when the specifications of the Helm charts change from one version to another? Dealing with these changes introduces the possibility of encountering breaking changes and invalid values. How we will deal within in the upgrades? 
- How do we handle breaking changes and invalid values? We need to develop robust mechanisms for validation and ensure a smooth transition in such cases.

Considering the challenges involved, this desire to accommodate any Helm chart value raises concerns. It is unclear whether and how this goal can be achieved in a maintainable and feasible manner. So, we need further investigation to determine if it is realistically feasible in the short and long term, as well as to understand the implications, in my humble opinion (IMHO).